File: lapras_agents/idle_task_detector_agent.py
import logging
import time
from typing import List, Optional
from dataclasses import dataclass
from datetime import datetime

from .agent import AgentComponent

logger = logging.getLogger(__name__)

@dataclass
class TaskNotification:
    task_type: str
    timestamp: int
    agent_name: str
    involved_agents: List[str]

class IdleTaskDetectorAgent(AgentComponent):
    PRESENCE_CHECK_INTERVAL = 60  # 1 minute in seconds
    
    def __init__(self, event_dispatcher, agent):
        super().__init__(event_dispatcher, agent)
        self.presence_history: List[str] = []
        self.current_presence: Optional[str] = None
        self.inferred_user_presence = None  # This would be set by the context system
        
    def subscribe_events(self):
        self.subscribe_event("CONTEXT_UPDATED")
        self.subscribe_event("TASK_NOTIFIED")  # For testing
        
    def handle_event(self, event) -> bool:
        if event.type == "CONTEXT_UPDATED":
            context = event.data
            if context.name == "InferredUserPresence":
                self.current_presence = context.value
                logger.info(f"Current presence: {self.current_presence}")
        return True
        
    def run(self):
        while True:
            if self.current_presence is not None:
                self.presence_history.append(self.current_presence)
                
            history_length = len(self.presence_history)
            
            if history_length > 10 and self.current_presence == "present":
                last_presence = self.presence_history[-10:]
                
                if all(p == "present" for p in last_presence):
                    self.notify_task_idle()
                    
            if history_length > 1000:
                self.presence_history.clear()
                
            time.sleep(self.PRESENCE_CHECK_INTERVAL)
            
    def notify_task_idle(self):
        logger.info(f"Presence history: {self.presence_history}")
        logger.info("Current task is Idle")
        
        timestamp = int(datetime.now().timestamp() * 1000)
        involved_agents = self.agent.config.get_option_as_array("involved_agents")
        
        task_notification = TaskNotification(
            task_type="Idle",
            timestamp=timestamp,
            agent_name=self.agent.config.agent_name,
            involved_agents=involved_agents
        )
        
        # This would be implemented based on your MQTT system
        # self.mqtt_communicator.publish(topic, task_notification) 

File: lapras-agents/src/main/java/kr/ac/kaist/cdsn/lapras/agents/IdleTaskDetectorAgent.java
package kr.ac.kaist.cdsn.lapras.agents;

import kr.ac.kaist.cdsn.lapras.Agent;
import kr.ac.kaist.cdsn.lapras.LaprasException;
import kr.ac.kaist.cdsn.lapras.agent.AgentComponent;
import kr.ac.kaist.cdsn.lapras.agent.Context;
import kr.ac.kaist.cdsn.lapras.communicator.LaprasTopic;
import kr.ac.kaist.cdsn.lapras.communicator.MessageType;
import kr.ac.kaist.cdsn.lapras.context.ContextField;
import kr.ac.kaist.cdsn.lapras.context.ContextInstance;
import kr.ac.kaist.cdsn.lapras.event.Event;
import kr.ac.kaist.cdsn.lapras.event.EventDispatcher;
import kr.ac.kaist.cdsn.lapras.event.EventType;
import kr.ac.kaist.cdsn.lapras.task.TaskNotification;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Created by JWP on 2017. 7. 10..
 */
public class IdleTaskDetectorAgent extends AgentComponent{
    private static final Logger LOGGER = LoggerFactory.getLogger(IdleTaskDetectorAgent.class);
    private static final long PRESENCE_CHECK_INTERVAL = 1 * 60 * 1000;

    private static List<String> presenseHistory = new ArrayList<>();
    private static String currentPresence = null;

    public IdleTaskDetectorAgent(EventDispatcher eventDispatcher, Agent agent) throws LaprasException {
        super(eventDispatcher, agent);
    }

    @ContextField
    private Context inferredUserPresence;

    @Override
    protected void subscribeEvents() {
        subscribeEvent(EventType.CONTEXT_UPDATED);
        subscribeEvent(EventType.TASK_NOTIFIED); // For Test
    }

    @Override
    protected boolean handleEvent(Event event) {
        switch (event.getType()) {
            case CONTEXT_UPDATED:
                ContextInstance context = (ContextInstance) event.getData();
                if (context.getName().equals("InferredUserPresence")) {
                    currentPresence = (String) context.getValue();
                    LOGGER.info("Current presence:" + currentPresence);
                }
        }
        return true;
    }

    /**
     * Check whether the current state is idle using the history of user presence.
     * If the state is idle, notify it to other agents to turn off themselves.
     */

    @Override
    public void run() {
        while (true) {
            if (currentPresence != null) {
                presenseHistory.add(currentPresence);
            }

            int historyLength = presenseHistory.size();

            if (historyLength > 10 && currentPresence.equals("present")) {
                List <String> lastPresence = presenseHistory.subList(Math.max(historyLength - 10, 0), historyLength);

                for (String element : lastPresence) {
                    if (!element.equals("present"))
                        break;

                    notifyTaskIdle();
                }
            }

            if (historyLength > 1000)
                presenseHistory.clear();

            try {
                Thread.sleep(PRESENCE_CHECK_INTERVAL);
            } catch (InterruptedException e) {
                break;
            }
        }
    }

    /**
     * Publish notifying message
     */

    private void notifyTaskIdle() {
        LOGGER.info(presenseHistory.toString());
        LOGGER.info("Current task is Idle");

        long timestamp = System.currentTimeMillis();
        List <String> involvedAgents = Arrays.asList(agent.getAgentConfig().getOptionAsArray("involved_agents"));

        LaprasTopic topic = new LaprasTopic(null, MessageType.TASK, "Idle");
        TaskNotification taskNotification = new TaskNotification("Idle", timestamp, agentName, involvedAgents);

        mqttCommunicator.publish(topic, taskNotification);
    }
}
